src/python/pants/backend/python/rules/prepare_chrooted_python_sources.py

```diff
@@ -0,0 +1,48 @@
+# Copyright 2020 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from dataclasses import dataclass
+
+from pants.backend.python.rules.inject_init import InjectedInitDigest
+from pants.engine.fs import Digest, DirectoriesToMerge
+from pants.engine.legacy.graph import HydratedTarget, HydratedTargets
+from pants.engine.rules import rule
+from pants.engine.selectors import Get, MultiGet
+from pants.rules.core.strip_source_root import SourceRootStrippedSources
+
+
+@dataclass(frozen=True)
+class ChrootedPythonSourcesRequest:
```

Comment:
There is a lot of the "Request" pattern in use right now. But one thing that I think would be good to consider would be biasing toward using return types as the way to differentiate rules: ie, have this rule return `ChrootedPythonSources` by consuming `HydratedTargets`.

Some reasons for this:
1. it allows for "overloading" of the way that something is computed: for example, perhaps `ChrootedPythonSources` could be computed given some other input.
2. it simplifies the rule graph: rather than having hundreds of ways to compute a Digest for various imports (another type of overloading), we have a rule graph that is more output-centric: a smaller number of rules that are able to compute `ChrootedPythonSources`.

Not a blocking comment: just a thought moving forward.